components/tabs/src/storage.rs

```diff
@@ -362,6 +372,34 @@ impl TabsStorage {
     }
 }
 
+// Trim the amount of tabs in a list to fit the specified memory size
+fn trim_tabs_length(
+    tabs: &mut Vec<RemoteTab>,
+    payload_size_max_bytes: usize,
+) -> Result<Vec<RemoteTab>> {
+    // Same as what desktop does in util.js
+    // See bug 535326 comment 8 for an explanation of the estimation
+    let max_serialized_size = (payload_size_max_bytes / 4) * 3 - 1500;
+    let size = compute_serialized_size(&tabs);
+    if size > max_serialized_size {
+        // Estimate a little more than the direct fraction to maximize packing
+        let cutoff = (tabs.len() * max_serialized_size) / size;
+        log::debug!("Truncating the tabs by: {}", cutoff);
+        tabs.truncate(cutoff);
+
+        // Keep dropping off the last entry until the data fits.
+        while compute_serialized_size(&tabs) > max_serialized_size {
+            tabs.pop();
+        }
+    }
+    Ok(tabs.to_vec())
+}
+
+fn compute_serialized_size(v: &Vec<RemoteTab>) -> usize {
+    let json = serde_json::to_string(v).unwrap();
+    mem::size_of_val(&*json)
```

Comment:
isn't this just `json.len()`?

Reply:
yeah from a quick test it looks like they're equivalent. I'll just use len here since it's simpler (might not even need to be a separate fn anymore, but i'll leave it as is)

```
[2023-02-15T21:36:30Z DEBUG tabs::storage] mem::size_of_val: 628768
[2023-02-15T21:36:30Z DEBUG tabs::storage] json.len(): 628768
```